### PR TITLE
Add OpenVINO export support for spark2_5 architecture (Spark-X2.5)

### LINKED ISSUE #2003: Add OpenVINO export support for spark2_5 architecture (Spark-X2.5-4B / 1.7B)
### Feature request

Please add OpenVINO export support for the `spark2_5` architecture (`Spark2_5ForCausalLM`), used by the Spark-X2.5 series.

- Model card: https://huggingface.co/XHToken/Spark-X2.5-4B (also a 1.7B variant)
- `model_type: spark2_5`, `architectures: ["Spark2_5ForCausalLM"]`

### Motivation

Spark-X2.5 is an on-device agentic model family with strong published benchmarks. We are evaluating it as a local replacement for our production model on **Intel Core Ultra (Meteor Lake) iGPU**, where **OpenVINO GenAI is our production inference stack**. Today there is no OpenVINO path for this architecture — neither `optimum-intel`, `openvino.genai`, nor `openvino` contains any reference to `spark2_5`, and it is not in the [OpenVINO GenAI supported-models list](https://openvinotoolkit.github.io/openvino.genai/docs/supported-models/).

### Architecture summary (from `config.json`)

Hybrid interleaved attention:

- **36 layers**, pattern = 3× sliding-window + 1× full attention, repeated ⇒ 27 sliding / 9 full
- `sliding_window: 512`
- GQA: `num_attention_heads: 16`, `num_key_value_heads: 4`, `head_dim: 256`
- `hidden_size: 2560`, `intermediate_size: 10240`
- `vocab_size: 131072`, `max_position_embeddings: 1048576`
- **Dual RoPE base** (NeoX-style): `10000` for sliding layers / `5000000` for full-attention layers
- **Gated attention**: `headwise_attn_output_gate: true`, `gate_attn_act_mode: sigmoid`
- `rms_norm_eps: 1e-06`

### Why explicit support is needed (not "similar architecture may also work")

The custom ops — the **headwise gated attention output gate** and the **interleaved SWA with dual RoPE base** — are exactly the parts that silently break on backends that have not implemented them. In our testing with llama.cpp (which added native `spark2_5` support in Sept 2026):

- CPU and the **OpenCL** GPU backend produce **correct** output.
- The **Vulkan** backend loads but produces **incoherent garbage** (numerically wrong attention) — both with and without flash-attention.
- The **Level-Zero (SYCL)** backend **segfaults at model load**.

So this architecture will not "just work" by architecture similarity — it needs explicit enablement.

### Ecosystem status

- **llama.cpp**: native `spark2_5` support added Sept 2026 (GGUF).
- **Ollama**: open request (ollama/ollama#18195).
- **OpenVINO / optimum-intel**: no code, no issue, not in the supported-models list.

### Ask

Is enabling `spark2_5` for OpenVINO export on the roadmap? We are happy to help validate on Intel Core Ultra (Meteor Lake) iGPU with OpenVINO GenAI once a conversion path exists.

### PR DESCRIPTION
**Pull request by bharqav**
# What does this PR do?

Fixes #2003

## Summary

This PR adds OpenVINO export support for the `spark2_5` architecture (Spark-X2.5-4B / 1.7B). This is a `trust_remote_code` model rather than a native `transformers` architecture: the Hugging Face repository ships its own `modeling_spark.py` and `configuration_spark.py`, resolved through `config.json`'s `auto_map`. The primary gap was registration in `optimum-intel`'s `TasksManager`, not a missing mapping to an existing `transformers` architecture.

## Architecture

`spark2_5` combines four characteristics that, in combination, have no exact precedent in the existing OpenVINO exporter configurations:

1. **Hybrid interleaved attention.** `layer_types` alternates three sliding-window layers with one full-attention layer (`sliding_window=512`), repeated across all 36 layers (27 sliding, 9 full). The reference implementation constructs this via `transformers.masking_utils.create_causal_mask` and `create_sliding_window_causal_mask`, selected per layer by `layer_type`. This follows the same pattern already used by `Gemma3TextOpenVINOConfig` / `Gemma4TextOpenVINOConfig`.
2. **Dual RoPE base with a per-layer-type partial rotary factor.** Sliding-window layers use `rope_theta=10000` with `partial_rotary_factor=1.0`; full-attention layers use `rope_theta=5000000` with `partial_rotary_factor=0.25`. This combination has no existing analogue elsewhere in the codebase and required direct numerical validation rather than confirmation by tracing alone.
3. **Headwise gated attention output**, via a per-head scalar gate (`g_proj`, sigmoid or silu) applied before the output projection.
4. **Grouped-query attention**, uniform across all layers: 16 query heads, 4 key/value heads.

## Implementation

Registration follows the existing `Gemma3Text` / `Llama` pattern in `optimum/exporters/openvino/model_configs.py`: a new `Spark25OpenVINOConfig(TextDecoderWithPositionIdsOpenVINOConfig)`, reusing `NormalizedTextConfig` (the reference config's field names already match what it expects, so no custom `NormalizedConfig` subclass was required), and the GQA-aware `MistralDummyPastKeyValuesGenerator`, consistent with `LlamaOpenVINOConfig`'s choice. No custom `_MODEL_PATCHER` was needed beyond the default `OVDecoderModelPatcher`.

## Validation

Correctness was established in stages against a locally built, random-init test fixture, rather than assumed from successful tracing alone:

- **Ablation to isolate the source of numerical error.** Three fixture variants were compared against their own PyTorch references: baseline, `partial_rotary_factor` forced identical across layer types, and headwise gating disabled. The measured difference remained effectively flat across all three (approximately 0.0052-0.0058 max absolute difference), ruling out both the dual-RoPE logic and the gating logic as the source of any discrepancy.
- **Identifying the actual source.** The flat ablation result pointed to OpenVINO's CPU inference precision, which defaults to `bfloat16` regardless of the fp32 export weight format. Forcing `INFERENCE_PRECISION_HINT=f32` reduced the difference to approximately 1e-7 (fp32 machine epsilon) at both 8- and 16-layer fixture sizes, including at `num_key_value_heads=4`, confirming the export is mathematically correct and that the residual difference under default settings is inference-precision noise rather than a defect in the export.
- **Deriving the test tolerance.** Measured default-precision differences grew with depth (0.0055 at 8 layers, 0.0078-0.0089 at 16 layers across repeated runs). A linear extrapolation to the real model's 36-layer depth, with a 2x safety margin, gives `atol=3e-2`, used for `spark2_5` in `test_decoder.py` and documented with this derivation and its limitations in a comment at that line. The real 36-layer checkpoint has not been measured directly.
- **End-to-end confirmation against a real Hub fixture.** A tiny random-init checkpoint reproducing the real architecture's `layer_types` interleave, `rope_parameters`, and gating, with the real tokenizer, is published at `bharqav/tiny-random-spark2_5` and registered in `HUB_MODEL_NAMES` and `REMOTE_CODE_MODELS` in `tests/openvino/utils_tests.py`. `test_compare_to_transformers` passes against it:
```
tests/openvino/test_decoder.py::OVModelForCausalLMIntegrationTest::test_compare_to_transformers_23_spark2_5 PASSED
```
- **Export & CLI Tests.** Added `spark2_5` to `tests/openvino/test_export.py` and `tests/openvino/test_exporters_cli.py` across all standard export tasks (`text-generation`, `text-generation-with-past`, `fp16`, `int8`, `tokenizers`). All passed:
```
tests/openvino/test_export.py::ExportModelTest::test_export_22_spark2_5 PASSED
tests/openvino/test_exporters_cli.py (9/9 passed: test_export, test_exporters_cli, fp16, int8, tokenizers)
```

  I am glad to have this checkpoint mirrored into `optimum-intel-internal-testing` if a maintainer prefers that over an external namespace.

## Test infrastructure fix (unrelated to spark2_5 specifically)

While adding spark2_5 to `test_export.py` and `test_exporters_cli.py`, I found that `_resolve_cached_model_paths` in `tests/openvino/utils_tests.py` breaks loading for any `REMOTE_CODE_MODELS` architecture once its checkpoint is present in the local HF cache: it substitutes the Hub repo ID with a local snapshot directory path, which changes how `transformers.dynamic_module_utils` resolves relative imports in remote code (it tries to resolve symlinked snapshot files against the underlying `blobs/` directory, where files have no `.py` extension, and fails). I reproduced this independently against `optimum-intel-internal-testing/tiny-random-minicpm` from a fresh download, unrelated to my own fixture, confirming it's a pre-existing bug rather than something specific to spark2_5. Fixed by excluding `REMOTE_CODE_MODELS` from the local-path substitution. Verified the full `test_decoder.py` suite (52 passed, 94 skipped via the standard `@slow` gate, 2 unrelated pre-existing failures traced to local cache state and a missing `diffusers` dependency, neither touched by this change).

## Documentation

Added `Spark-X2.5` to the supported architectures list in `docs/source/openvino/models.mdx`.

## Pre-existing issues identified in the reference implementation

While constructing the test fixture, two issues were found in the model's published `modeling_spark.py` / `configuration_spark.py`. These are unrelated to this PR and are not fixed here; noted for visibility:

- `_tied_weights_keys` is defined as a list (`["lm_head.weight"]`), consistent with transformers 4.x conventions. Under transformers 5.x, `get_expanded_tied_weights_keys` expects a dict mapping and raises `AttributeError` otherwise.
- The `mask_kwargs` dict passed to `create_causal_mask` / `create_sliding_window_causal_mask` uses `input_embeds` and `cache_position`, neither of which the current `transformers.masking_utils` signature accepts.

Both were addressed locally, within the test fixture only. The upstream Hub repository was not modified.

## Scope

Not covered in this PR: OpenVINO GenAI's separate supported-models list (a different repository) and quantization recipes.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: tests/openvino/test_decoder.py [hunk 1]
```diff
@@ -399,7 +400,15 @@ def test_compare_to_transformers(self, model_arch):
...
+        # spark2_5: hybrid interleaved sliding/full attention + dual RoPE base, 36 layers at real model scale.
+        # Forced-fp32 export verified mathematically exact (~1e-7) on 8- and 16-layer test fixtures; this atol
+        # covers default bf16 CPU inference precision, extrapolated from measured 8L/16L diffs (0.0055/0.0078)
+        # to 36L (~0.0136) with a 2x safety margin, since we have not measured the real 36-layer checkpoint directly.
+        atol = (
+            3e-2
+            if model_arch == "spark2_5"
+            else (3e-3 if model_arch in ["minicpm", "qwen2-moe", "gemma3n_text"] else 1e-4)
+        )
         # quantized models have different logits value range
```
Comment (rkazants) [inline tests/openvino/test_decoder.py:412]: quite strange, smth is missed

File: tests/openvino/test_decoder.py
Comment (rkazants) [inline tests/openvino/test_decoder.py:1]: please provide export tests and the same set of tests we usually provide for text-generation models

Comment (bharqav) [reply tests/openvino/test_decoder.py:1]: Added spark2_5 to test_export.py (SUPPORTED_ARCHITECTURES) and test_exporters_cli.py (SUPPORTED_ARCHITECTURES, EXPECTED_NUMBER_OF_TOKENIZER_MODELS, _ARCHITECTURES_TO_EXPECTED_INT8), matching the existing pattern for other text-generation architectures. All passing locally. Also found and fixed a pre-existing bug in _resolve_cached_model_paths that this surfaced, detailed in the PR description.
